### PR TITLE
Added naming convention section to the README

### PR DESCRIPTION
**Pull request by ThibaultKlein**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
 
 ## Standards ##
 
+### Naming Convention ###
+
+With Swift 3, Apple updated the Swift [API design guidelines](https://swift.org/documentation/api-design-guidelines/) to promote clarity over brevity.
+
+Capitalize all type names with camel case, while function and variable names start with a lower case. Avoid acronyms and abbreviations for clarity and readability.
```
Comment (paulmiard) [inline README.md]: This is not easy to read, you talk about type names and functions and variables in the same sentence.. Maybe a bunch of bullet points + examples would help?

File: README.md [hunk 2]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+
+#### Protocols ####
+
+Protocols describing something should be a noun: `Collection`, `Element`. Protocols describing an ability should end with “ing” or “able”: `Evaluatable`, `Printable`, `Formatting`.
```
Comment (paulmiard) [inline README.md]: I don't get what the first sentence means. `A protocol should BE a noun`? Could be more clear and maybe rephrase? Also use `for example` or `e.g.`

File: README.md [hunk 3]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+
+#### Enum ####
+
+Enum cases should start with a lower case and be camel case.
```
Comment (paulmiard) [inline README.md]: start with a lowercase letter and use camelCase

OR

start with a lower case letter and use camel case

Comment (paulmiard) [reply README.md]: Also show a camelCase example in your example

File: README.md [hunk 4]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+enum Color {
+    case red
+    case blue
+    case green
+}
+```
+
+#### Functions ####
+
+Include words that can make the function’s behavior clearer. Example of a function that removes an element at an index x.
```
Comment (paulmiard) [inline README.md]: Not clear..

How about `You should name your function with words that describe its behavior. Let's take the example of a function that removes and element at a given index 'x':`

File: README.md [hunk 5]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+```
+
+**Not Preferred:**
+```swift
+func remove(index: Index) -> Element
+```
+
+*Rationale*: It is better to specify that we are removing the element at the given index, and we are not trying to remove the given parameter itself, to make the behavior of the function very clear.
+
+Avoid unnecessary words in the function name.
```
Comment (paulmiard) [inline README.md:117]: is it part of the rationale? Is it another rule? Looks lost..

File: README.md [hunk 6]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+}
+```
+
+#### Functions ####
+
+Include words that can make the function’s behavior clearer. Example of a function that removes an element at an index x.
+
+**Preferred:**
+```swift
+func remove(at index: Index) -> Element
```
Comment (paulmiard) [inline README.md:107]: Index -> Int ?

File: README.md (hunk 5 above)
Comment (paulmiard) [reply README.md:117]: Oh I get it there is an example after. Well I need to check how it looks when laid out

File: README.md [hunk 7]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+```swift
+func remove(_ element: Element) -> Element?
+```
+
+**Not Preferred:**
+```swift
+func removeElement(_ element: Element) -> Element?
+```
+
+*Rationale*: Make the code clear and concise whenever possible. Adding extra unnecessary words will make the code harder to read and understand.
```
Comment (paulmiard) [inline README.md]: Make -> `You should make`

File: README.md [hunk 8]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
 
 ## Standards ##
 
+### Naming Convention ###
+
+With Swift 3, Apple updated the Swift [API design guidelines](https://swift.org/documentation/api-design-guidelines/) to promote clarity over brevity.
```
Comment (paulmiard) [inline README.md]: We probably don't won't to mention Swift 3 anywhere.. Or once at the very beginning to explain our guidelines are Swift3+

File: README.md [hunk 9]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+```
+
+**Not Preferred:**
+```swift
+func removeElement(_ element: Element) -> Element?
+```
+
+*Rationale*: Make the code clear and concise whenever possible. Adding extra unnecessary words will make the code harder to read and understand.
+
+Swift 3 encourages to name functions to be read as a sentence according to their side effects.
```
Comment (paulmiard) [inline README.md]: Try to remove reference to Swift 3

File: README.md [hunk 10]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+
+**Preferred:**
+```swift
+x.insert(y, at: z)          “x, insert y at z”
+x.subViews(havingColor: y)  “x's subviews having color y”
+x.capitalizingNouns()       “x, capitalizing nouns”
```
Comment (paulmiard) [inline README.md]: This block won't compile. Let's use comment style to explain what it does?

File: README.md [hunk 11]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+```
+
+**Not Preferred:**
+```swift
+x.insert(y, position: z)
+x.subViews(color: y)
+x.nounCapitalize()
```
Comment (paulmiard) [inline README.md:159]: wouldn't the not preferred example to the previous one be `x.nounCapitalizing()` ?

File: README.md [hunk 12]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+```
+
+* With side effects: use **imperative verb** phrases
+	* `print(x)`, `x.sort()`, `x.append(y)`
+* Without side effects: use **noun** phrases
+	* `x.formattedName()`, `x.successor()`
```
Comment (paulmiard) [inline README.md]: Can we move that before the preferred/not prefered?

File: README.md [hunk 13]
```diff
@@ -59,6 +59,88 @@ style guide that you will not find in the GitHub one.
...
+
+When the function can be described by a verb, use an imperative verb for the mutating function and apply “ed” or “ing” to the nonmutating function:
```
Comment (paulmiard) [inline README.md]: nonmutating -> non-mutating (there is a keyword  nonmutating but here it's used in a sentence in English)..

Comment (paulmiard) [reply README.md]: also let's move that before the  preferred/not prefered as well

**Review by paulmiard — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (ThibaultKlein) [reply README.md:107]: no Index is the proper type for that

File: README.md (hunk 4 above)
Comment (ThibaultKlein) [reply README.md]: We shouldn't say "You should" here, I suggest "Name your function with words that describe its behavior"

File: README.md (hunk 7 above)
Comment (ThibaultKlein) [reply README.md]: Again let's try to avoid should as it is a guide and not recommendations

File: README.md (hunk 11 above)
Comment (ThibaultKlein) [reply README.md:159]: `nounCapitalize` is also a function name we want to avoid

File: README.md (hunk 10 above)
Comment (ThibaultKlein) [reply README.md]: capitalizingNouns is a dummy example here, I don't think we need to make it compile necesseraly

File: README.md (hunk 13 above)
Comment (ThibaultKlein) [reply README.md]: it seems that [nonmutating](https://en.wiktionary.org/wiki/nonmutating) is the correct way

**Comment by ThibaultKlein** [comment]
@paulmiard please review again, I updated the pull request

File: README.md [hunk 14]
```diff
@@ -59,6 +59,103 @@ style guide that you will not find in the GitHub one.
 
 ## Standards ##
 
+### Naming Convention ###
```
Comment (htinlinn) [inline README.md]: Should this be pluralized? "Naming Conventions"?

File: README.md [hunk 15]
```diff
@@ -59,6 +59,103 @@ style guide that you will not find in the GitHub one.
...
+
+The Apple Swift [API design guidelines](https://swift.org/documentation/api-design-guidelines/) promote clarity over brevity.
+
+Capitalize all type names with camel case, while function and variable names start with a lower case.
+
+```swift
+class MyClass { }
+protocol MyProtocol { }
+func myFunction { }
```
Comment (htinlinn) [inline README.md]: This should be `func myFunction() { }`. Function definition requires parentheses.

File: README.md [hunk 16]
```diff
@@ -59,6 +59,103 @@ style guide that you will not find in the GitHub one.
 
 ## Standards ##
 
+### Naming Convention ###
+
+The Apple Swift [API design guidelines](https://swift.org/documentation/api-design-guidelines/) promote clarity over brevity.
```
Comment (htinlinn) [inline README.md]: This sentence here feels out of place to me, especially because we don't elaborate on what "clarity over brevity" is and why it matters. Maybe this should just say "Apple's Swift API design guidelines promote clarity above everything else" or something along that line?

File: README.md [hunk 17]
```diff
@@ -59,6 +59,103 @@ style guide that you will not find in the GitHub one.
...
+
+When the function can be described by a verb, use an imperative verb for the mutating function and apply “ed” or “ing” to the nonmutating function:
+* Mutating function
+	* `x.sort()`
+	* `x.append(y)`
+* Nonmutating function
+	* `z = x.sorted()`
+	* `z = x.appending(y)`
+
+Apple encourages to name functions to be read as a sentence according to their side effects.
```
Comment (htinlinn) [inline README.md]: Just nitpicking here. But, I think we should drop the Apple part and make this sentence sound more like a command like the sentences before this.

File: README.md (hunk 15 above)
Comment (ThibaultKlein) [reply README.md]: done

File: README.md (hunk 14 above)
Comment (ThibaultKlein) [reply README.md]: done

File: README.md (hunk 16 above)
Comment (ThibaultKlein) [reply README.md]: Added

> Code should be concise, readable and clear at the point of use. However, having compact code that sacrifices clarity is a non-goal.

File: README.md (hunk 12 above)
Comment (ThibaultKlein) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (ThibaultKlein) [reply README.md]: done

File: README.md (hunk 9 above)
Comment (ThibaultKlein) [reply README.md]: done

File: README.md (hunk 17 above)
Comment (ThibaultKlein) [reply README.md]: agreed, done

File: README.md [hunk 18]
```diff
@@ -59,6 +59,105 @@ style guide that you will not find in the GitHub one.
...
+userID = "12345"
+```
+
+#### Protocols ####
+
+Protocol names describing something should be a noun: `Collection`, `Element`. Protocol names describing an ability should end with “ing” or “able”: `Evaluatable`, `Printable`, `Formatting`.
+
+#### Enums ####
+
+Enum cases start with a lower case and use camel case.
```
Comment (paulmiard) [inline README.md]: `a lower case` -> `a lower case letter`

File: README.md [hunk 19]
```diff
@@ -59,6 +59,105 @@ style guide that you will not find in the GitHub one.
...
+The Apple Swift [API design guidelines](https://swift.org/documentation/api-design-guidelines/) promote clarity over brevity. Code should be concise, readable and clear at the point of use. However, having compact code that sacrifices clarity is a non-goal.
+
+#### Types ####
+
+Capitalize all type names with camel case, while function and variable names start with a lower case.
+
+```swift
+class MyClass { }
+protocol MyProtocol { }
+func myFunction() { }
```
Comment (paulmiard) [inline README.md:73]: add a variable example?

File: README.md [hunk 20]
```diff
@@ -59,6 +59,105 @@ style guide that you will not find in the GitHub one.
 
 ## Standards ##
 
+### Naming Conventions ###
+
+The Apple Swift [API design guidelines](https://swift.org/documentation/api-design-guidelines/) promote clarity over brevity. Code should be concise, readable and clear at the point of use. However, having compact code that sacrifices clarity is a non-goal.
+
+#### Types ####
+
+Capitalize all type names with camel case, while function and variable names start with a lower case.
```
Comment (paulmiard) [inline README.md]: Apple in their guidelines: `Follow case conventions. Names of types and protocols are UpperCamelCase. Everything else is lowerCamelCase.`

Let's copy that?

File: README.md (hunk 18 above)
Comment (paulmiard) [reply README.md]: I checked the output and it's hard to know we are still in functions section.. This is a second bullet in functuons, then there is a third one.. Maybe we can style that differently to make it clear what we are reading?

Comment (paulmiard) [reply README.md]: Linn and I checked and looks like we could bump every section's number of `#` down in the entire file. Should be a different PR, but we can still look into that susub section case here, comment will still be valid. Maybe use bullet points?

File: README.md [hunk 21]
```diff
@@ -59,6 +59,105 @@ style guide that you will not find in the GitHub one.
...
+```swift
+func remove(_ element: Element) -> Element?
+```
+
+**Not Preferred:**
+```swift
+func removeElement(_ element: Element) -> Element?
+```
+
+*Rationale*: It makes the code clear and more concise. Adding extra unnecessary words will make the code harder to read and understand.
```
Comment (paulmiard) [inline README.md]: `clear` -> `clearer` (to be consistent with `more concise`)

File: README.md [hunk 22]
```diff
@@ -59,6 +59,105 @@ style guide that you will not find in the GitHub one.
...
+
+Name your functions based on their side effects and behaviors.
+
+* With side effects: use **imperative verb** phrases:
+	* `print(x)`, `x.sort()`, `x.append(y)`
+* Without side effects: use **noun** phrases:
+	* `x.formattedName()`, `x.successor()`
+
+When the function can be described by a verb, use an imperative verb for the mutating function and apply “ed” or “ing” to the nonmutating function:
```
Comment (paulmiard) [inline README.md:138]: `described by` -> `described with`?? Not sure about this one..

File: README.md [hunk 23]
```diff
@@ -59,6 +59,105 @@ style guide that you will not find in the GitHub one.
...
+* Mutating function:
+	* `x.sort()`
+	* `x.append(y)`
+* Nonmutating function:
+	* `z = x.sorted()`
+	* `z = x.appending(y)`
+
+Name functions to be read as a sentence according to their side effects.
```
Comment (paulmiard) [inline README.md:146]: Once you add the bullet points (see previous comment) you should keep each consistent (eg `name functions` versus `name your functions`)

File: README.md (hunk 19 above)
Comment (paulmiard) [reply README.md:73]: Talking about the “x, capitalizing nouns” part. And the other comments. You should add `//` so Github uses the right color scheme

File: README.md (hunk 20 above)
Comment (paulmiard) [reply README.md]: Yes if you want to mirror the previous block don't you wamt to use `nounCapitalizing` ? And that would be wrong as well..

File: README.md [hunk 24]
```diff
@@ -518,7 +616,7 @@ of how they wish to handle the result of this failable operation.
 
 In general, you should avoid `NSError` in Swift in favor of defining your own `ErrorType`. However, in the event you do need to use `NSError` (for interop with Objective-C, for instance):
 
-* Define a proper domain for your `NSError`. This should be specific to your module and ideally would be reflective of your bundle identifier (i.e. `com.prolificinteractive.MyApp`)
+* Define a proper domain for your `NSError`. This should be specific to your module and ideally would be reflective of your bundle identifier (i.e. `com.prolificinteractive.MyApp`).
```
Comment (paulmiard) [inline README.md]: `i.e.` -> `e.g.`

**Review by paulmiard — CHANGES_REQUESTED** [review]

File: README.md (hunk 20 above)
Comment (ThibaultKlein) [reply README.md]: I changed this part with `Write all type names with UpperCamelCase, function and variable names with lowerCamelCase.`

File: README.md (hunk 18 above)
Comment (ThibaultKlein) [reply README.md]: I changed to `Enum cases start with lowerCamelCase.`

Not a fan of bullet points here, would make the guide less readable IMO. Enum isn't in function section but in Naming conventions section. Let me know your thoughts about the file structure.

File: README.md (hunk 21 above)
Comment (ThibaultKlein) [reply README.md]: Updated.

File: README.md (hunk 23 above)
Comment (ThibaultKlein) [reply README.md:146]: Not doing the bullet points formatting here. Let's review that in a separate PR.

File: README.md (hunk 19 above)
Comment (ThibaultKlein) [reply README.md:73]: Updated.

File: README.md (hunk 22 above)
Comment (ThibaultKlein) [reply README.md:138]: Based on [this](http://english.stackexchange.com/a/26629), if the sentence has a passive voice, described by is the preferred option. If the sentence has an active voice, described with is preferred.

File: README.md (hunk 10 above)
Comment (paulmiard) [reply README.md]: Well that's matter of consistency throughout the guidelines. The right side of those three examples should use the comment notation imo, but it's not blocking

File: README.md (hunk 11 above)
Comment (paulmiard) [reply README.md:159]: I get that, was just wondering if we want to mirror the _preferred_ section. Since you used `capitalizingNouns` I thought `nounCapitalizing` would be a better counter example. Not blocking though

**Review by paulmiard — APPROVED** [review]

File: README.md (hunk 23 above)
Comment (paulmiard) [reply README.md:146]: We should make a note of that

Comment (ThibaultKlein) [reply README.md:146]: Added a note in our discussion document

File: README.md (hunk 10 above)
Comment (ThibaultKlein) [reply README.md]: I forgot to push my update for this section of the readme, it should be updated now

File: README.md (hunk 11 above)
Comment (ThibaultKlein) [reply README.md:159]: The goal here is to also show that `nounCapitalize` doesn't sound english which you should avoid. If you prefer to use `nounCapitalizing` I can update

Comment (htinlinn) [reply README.md:159]: This is an example that Apple also uses so let's just stick with `nounCapitalize()`...

**Review by htinlinn — APPROVED** [review]
